pkg/ddl/partition.go

```diff
@@ -2872,6 +2936,33 @@ func (w *worker) onReorganizePartition(d *ddlCtx, t *meta.Meta, job *model.Job)
 			}
 		}
 
+		// All global indexes must be recreated, we cannot update them in-place, since we must have
+		// both old and new set of partition ids in the unique index at the same time!
+		for _, index := range tblInfo.Indices {
+			if !index.Unique {
+				// for now, only unique index can be global, non-unique indexes are 'local'
+				continue
+			}
+			inAllPartitionColumns, err := checkPartitionKeysConstraint(partInfo, index.Columns, tblInfo)
+			if err != nil {
+				return ver, errors.Trace(err)
+			}
+			if index.Global || !inAllPartitionColumns {
+				// Duplicate the unique indexes with new index ids.
+				// If previously was Global or will be Global:
+				// it must be recreated with new index ID
+				newIndex := index.Clone()
+				newIndex.State = model.StateDeleteOnly
+				newIndex.ID = AllocateIndexID(tblInfo)
+				if inAllPartitionColumns {
+					newIndex.Global = false
+				} else {
+					// If not including all partitioning columns, make it Global
+					newIndex.Global = true
+				}
+				tblInfo.Indices = append(tblInfo.Indices, newIndex)
```

Comment:
It seems there is a chance to have two "primary keys", I am not sure if there are problems

```sql
CREATE TABLE `t3` (
  `a` int(11) NOT NULL,
  `b` int(11) DEFAULT NULL,
  PRIMARY KEY (`a`) /*T![clustered_index] NONCLUSTERED */,
  UNIQUE KEY `idx_b` (`b`),
  UNIQUE KEY `idx_a` (`a`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
PARTITION BY HASH (`a`) PARTITIONS 3

alter table t3 partition by hash(b) partitions 3;
```

Reply:
Maybe we can add DML tests during reorganizing partitions in next PR.

Reply:
There should only be one 'PRIMARY KEY' (due to definition of what a primary key is) but multiple candidate keys are possible (i.e. unique indexes of not null columns), both using global index as well as non-partitioned tables.
Also I don't think there are any limitations for having duplicate indexes with different names, are there?

Reply:
> Also I don't think there are any limitations for having duplicate indexes with different names, are there?

+1

I mean there are multiple indexes with `index.Primary == true`, it breaks the assumption of "there is only one primary index at anytime". I can't figure out any scenarios that can cause errors for now